src/qt/governancelist.cpp

```diff
@@ -428,3 +466,158 @@ void GovernanceList::showAdditionalInfo(const QModelIndex& index)
+        if (walletModel->wallet().isSpendable(script)) {
+            votableMasternodes[dmn.proTxHash] = dmn.pdmnState->keyIDVoting;
+        }
+    });
+
+    // Update masternode count display
+    updateMasternodeCount();
+}
+
+void GovernanceList::updateMasternodeCount() const
+{
+    if (ui && ui->mnCountLabel) {
+        ui->mnCountLabel->setText(QString::number(votableMasternodes.size()));
+    }
+}
+
+void GovernanceList::voteYes() { voteForProposal(VOTE_OUTCOME_YES); }
+
+void GovernanceList::voteNo() { voteForProposal(VOTE_OUTCOME_NO); }
+
+void GovernanceList::voteAbstain() { voteForProposal(VOTE_OUTCOME_ABSTAIN); }
+
+void GovernanceList::voteForProposal(vote_outcome_enum_t outcome)
+{
+    if (!walletModel) {
+        QMessageBox::warning(this, tr("Voting Failed"), tr("No wallet available."));
+        return;
+    }
+
+    if (votableMasternodes.empty()) {
+        QMessageBox::warning(this, tr("Voting Failed"), tr("No masternode voting keys found in wallet."));
+        return;
+    }
+
+    // Get the selected proposal
+    const auto selection = ui->govTableView->selectionModel()->selectedRows();
+    if (selection.isEmpty()) {
+        QMessageBox::warning(this, tr("Voting Failed"), tr("Please select a proposal to vote on."));
+        return;
+    }
+
+    const auto index = selection.first();
+    const auto proposal = proposalModel->getProposalAt(proposalModelProxy->mapToSource(index));
+    if (proposal == nullptr) return;
+
+    const uint256 proposalHash(uint256S(proposal->hash().toStdString()));
+
+    // Request unlock if needed and keep context alive for the voting operation
+    WalletModel::UnlockContext ctx(walletModel->requestUnlock());
+    if (!ctx.isValid()) {
+        // Unlock cancelled or failed
+        QMessageBox::warning(this, tr("Voting Failed"), tr("Unable to unlock wallet."));
+        return;
+    }
+
+    int nSuccessful = 0;
+    int nFailed = 0;
+    QStringList failedMessages;
+
+    // Get masternode list once before the loop
+    auto [mnList, pindex] = clientModel->getMasternodeList();
+    if (!pindex) {
+        QMessageBox::warning(this, tr("Voting Failed"), tr("Unable to get masternode list. Please try again later."));
+        return;
+    }
+
+    // Vote with each masternode
+    for (const auto& [proTxHash, votingKeyID] : votableMasternodes) {
+        // Find the masternode
+        auto dmn = mnList.GetValidMN(proTxHash);
+        if (!dmn) {
+            nFailed++;
+            failedMessages.append(tr("Masternode %1 not found").arg(QString::fromStdString(proTxHash.ToString())));
+            continue;
+        }
+
+        // Create vote
+        CGovernanceVote vote(dmn->collateralOutpoint, proposalHash, VOTE_SIGNAL_FUNDING, outcome);
+
+        // Sign vote
+        bool signSuccess = false;
+
+        // Special implementation for testnet (same as RPC SignVote)
+        if (Params().NetworkIDString() == CBaseChainParams::TESTNET) {
```

Comment:
let's re-use helper from rpc code here?
```
static bool SignVote(const CWallet& wallet, const CKeyID& keyID, CGovernanceVote& vote)
```

Reply:
see https://github.com/dashpay/dash/pull/6690/commits/20effe0d9497e505e7a8ac7c14fa28e5632fb223, moving to src/governance causes circular dependency